eip/EIPS/eip-4337.md

```diff
@@ -98,45 +97,66 @@ The wallet
+  external;
+
   function getAggregator() view returns (address);
 }
 ```
 * **getAggregator()** returns the aggregator this wallet supports.
-* The wallet is guaranteed to that the EntryPoint will only call its `validateUserOp` (with no signature) after calling the aggregator's validateSignatures
+* **validateUserOp()** MUST verify the `aggregator` parameter is valid and the same as `getAggregator`
 * The wallet should also support aggregator-specific getter (e.g. `getAggregationInfo()`).
   This method should export the wallet's public-key to the aggregator, and possibly more info
-* validateUserOp shouldn't validate the signature (since it is verified by the aggregator),
-  but it MUST verify the signature field is shorter than 63 bytes
+  (note that it is not called directly by the entryPoint) 
+* validateUserOp MAY completely ignore the signature field
 
 The core interface required by an aggregator is:
 ```solidity
 interface IAggregator {
 
-  function validateSignatures(address entryPoint, UserOperation[] calldata userOps, bytes calldata signature) view external;
+  function validateUserOpSignature(UserOperation calldata userOp) external view returns (bytes memory sigForUserOp, bytes memory sigForAggregation);
+
+  function aggregateSignatures(bytes[] calldata sigsForAggregation) external view returns (bytes memory aggregatesSignature);
+
+  function validateSignatures(UserOperation[] calldata userOps, bytes calldata signature) view external;
 }
 ```
 
-* bundlers/nodes must have explicit support for each aggregator, and know the code they need to calculate its aggregated signature.
-* the validateSignatures method is called by the entrypoint to let the aggregator validate the aggregate signature is valid for all UserOps.
-* if the signature is invalid or doesn't match, the aggregator should revert.
-* the aggregator SHOULDN'T call the getAggregator() of each wallet: the entrypoint already validated that on each userOp before calling validateSignatures
+* **validateUserOpSignature()** must validate the userOp's signature against the UserOp's hash (the same hash MUST be used for
+  **validateUserOpSignature** and later for **validateSignatures**)
+* it is called by the bunder just after simulateValidation() returns the aggregator's address.
+* The method should return a replacement value for the UserOp.signature, and a value used for aggregation
+  (the trivial "split" is return "" for the sigForUserOp, and the signature itself for sigForAggregation) 
+* **aggregateSignatures()** must aggregate all "sigForAggregation" into a single value.
+* **validateSignatures()** MUST validate the aggregated signature matches for all UserOperations in the array, and revert otherwise.
+  This method is called on-chain by `handleOps()`
+
+#### Trusting aggregators and off-chain optimization
+
+The aggregators SHOULD stake just like a paymaster. Bundlers MAY throttle down and ban aggregators in case they take too much 
```

Comment:
aggregator can grief a bundler, e.g. by wasting cpu while calculating signature. by requiring a stake, it can be throttled down, so it can do it at most once a day or so.
(at first stage we're going to use whitelisted aggregators, so it is less important to stake)